torch/nn/modules/container.py

```diff
@@ -234,7 +234,10 @@ def extend(self, sequential) -> 'Sequential':
         return self
 
 
-class ModuleList(Module):
+# `Sequence` but not a `MutableSequence` since as currently implemented, the signatures
+# of the mutable methods are incompatible, e.g. `append` returns `self` and `pop` takes
+# a slice.
+class ModuleList(Module, Sequence[T]):
```

Comment:
What is the impact of changing the inheritance to also include Sequence here?
In particular, this now doesn't have straight line inheritance, can that cause issues (especially due to how we use super())?
Are there other methods on the Sequence class that are not defined here that would change behavior?

Reply:
This will cause `super()` within `Module` when called from `ModuleList`'s call to `super()` to return `Sequence`, not `Object`. This should be fine and proper. `Module`'s only use of super is to call the un-modified get/setattr, and Sequence doesn't override that, so it will fall back to Object and be in practice a noop in resolution. But regardless of what super calls Module is making, even if it adds some later on, unless Sequence starts overriding something in Object, this will be a noop.

Another way to put it: the Method Resolution Order of Module remains unchanged with this change, and the MRO of ModuleList goes from [ModuleList, Module, Object] to [ModuleList, Module, Sequence, Object]

https://stackoverflow.com/a/3277407 was the reference I used most while trying to confirm this. 



Re: other methods. https://docs.python.org/3/library/collections.abc.html#collections-abstract-base-classes lists the abstract methods (that must be implemented) and the mixins (that can be, or can be derived from the abstract methods) for Sequence. The following are unimplemented by ModuleList: `__contains__`, `__reversed__`, `index`, and `count`. These are implemented as one would expect for a bog-standard default implementation: https://github.com/python/cpython/blob/a834e2d8e1230c17193c19b425e83e0bf736179e/Lib/_collections_abc.py#L1025

So it will grow those additional methods, but not change behavior of any existing methods. If it is problematic to expose these new ones, we can get the generics without the change to conform to the Sequence API by subclassing `Generic[T]` instead of `Sequence[T]`, but given that this really is exposing the Sequence interface, it seems proper to actually subclass it so that functions that take Sequences operate correctly.

Reply:
It's mostly the typing fixes. I've come across two separate implementations in the wild, and written my own, of a TypedModuleList that just wraps ModuleList and exists solely to get proper static types for homogenous ModuleLists. My goal was to just stop people from writing those.

In theory it also causes `__contains__`, `__reversed__`, and `index` to get exposed, but I'm not sure anyone cares about any of those enough to consider that an upside.

Reply:
I don't believe so. At the very least, in order to get the valuable part of this, you need `ModuleList` to be generic, and the only way to get generics is to subclass Generic[T] (https://docs.python.org/3/library/typing.html#user-defined-generic-types)

I was trying to see whether structural subtyping let us get the typing behavior without the subclassing, but today I learned that `isinstance` actually has magic based on structural subtyping such that, even if Sequence used structural subtyping (it doesn't), the very act of implementing its structure would cause `isinstance` to start returning True!

For example: https://docs.python.org/3/library/collections.abc.html#collections.abc.Iterable says:
> Checking isinstance(obj, Iterable) detects classes that are registered as [Iterable](https://docs.python.org/3/library/collections.abc.html#collections.abc.Iterable) or that have an __iter__() method
which I super didn't realize

So like, I guess someone in theory could be checking `isinstance(Sequence)` but that feels like getting eerily close to saying someone could check `hasattr(Module, 'spoooky')` so adding methods is breaking BC

Reply:
For my own future reference as much as anything, I'm now testing this (locally) with 
```
# Uses https://github.com/davidfritzsche/pytest-mypy-testing which must be installed, otherwise this silently fails to run
@pytest.mark.mypy_testing
def mypy_test_modulelist() -> None:
    ml = torch.nn.ModuleList([nn.Linear(10, 10), nn.Linear(10, 10)])
    reveal_type(ml)  # R: torch.nn.modules.container.ModuleList[torch.nn.modules.linear.Linear]
    # Make sure we know the type of fields on elements in the ModuleList.
    reveal_type(ml[0].weight)  # R: torch._tensor.Tensor
```

which does in fact fail at HEAD and passes with this change. That testing plugin is a bit funky though so I'm not even gonna try and land the test. FWIW it also does confirm that this passes when replacing Sequence[T] with Generic[T], but as above, I think saying either of those would be BC-breaking is a stretch unless I'm missing some other potential interaction.